File: algorithms/[1423]可获得的最大点数/solution.ts
// topics = ["数组"]

/**
 * 枚举所有可能的组合
 * time O(n), space O(n), n 为数组 cardPoints 的长度
 */
function maxScore(cardPoints: number[], k: number): number {
  const n = cardPoints.length;
  // 初始选取下标 0 ~ k - 1
  const stack = cardPoints.slice(0, k);
  let sum = stack.reduce((a, b) => a + b, 0);
  let res = sum;

  let i = n - 1;
  while (stack.length) {
    sum += cardPoints[i] - stack.pop()!;
    res = Math.max(res, sum);
    i -= 1;
  }

  return res;
}

File: algorithms/[1423]可获得的最大点数/solution.py
# topics = ["数组", "滑动窗口"]

from typing import List


class Solution:
    def maxScore(self, cardPoints: List[int], k: int) -> int:
        """
        由于只能从开头和末尾拿 k 张卡牌，所以最后剩下的必然是连续的 n-k 张卡牌，问题便转化为求出剩余卡牌点数之和的最小值
        time O(n), space O(1), n 为数组 cardPoints 的长度
        """
        n = len(cardPoints)
        min_value = total = sum(cardPoints[: n - k])
        for i in range(1, k + 1):
            total += cardPoints[n - k + i - 1] - cardPoints[i - 1]
            min_value = min(min_value, total)

        return sum(cardPoints) - min_value
